File: C++/CYCFactoring.cpp
// reference:
// H.M.Edwards "Fermat's Last Theorem" sections 4.4, 4.7

// uses NTL
//   http://www.shoup.net/ntl

#include "CYCFactoring.h"
#include "ZZFactoring.h"
#include "Period.h"
#include<NTL/ZZ_pXFactoring.h>
#include<NTL/LLL.h>
#include<exception>
using namespace NTL;

long SolCongr(Vec<ZZ_p>& u, const ZZ_p& u0)
// given u0 == eta_0 (mod p)
// get u_j == eta_j (mod p) (j=1,...,e-1)
// and output u_j (j=0,...,e-1)
// return 0 if successful, else return -1
{
    long i,j,e1(Period::e-1);
    ZZ_p t;
    Mat<ZZ_p> A;
    Vec<ZZ_p> v;
    A.SetDims(e1,e1);
    v.SetLength(e1);
    for(i=0; i<e1; i++) {
        for(j=0; j<e1; j++)
            A[j][i] = Period::W[i+1][j+1];
        A[i][i] -= u0;
        mul(v[i], u0, -Period::W[i+1][0]);
    }
    solve(t,v,A,v);
    if(IsZero(t)) return -1;
    u.SetLength(Period::e);
    u[0] = u0;
    for(i=0; i<e1; i++) u[i+1] = v[i];
    return 0;
}

void FactorPrime(CYC& q, const ZZ& p, long f)
// q = cyclotomic prime integer that divides prime number p
// f = order of p mod CYC::n
// Assume CYC::init(n) has been executed
// If f<=0, f is set to CYC::ord(p)
{
    if(p == CYC::n) { q=1; SetCoeff(q,1,-1); return; }
    if(f<=0) f = CYC::ord(p);
    if(f == CYC::n-1) { q=p; return; }

    long i,j,e((CYC::n-1)/f),e1(e-1);
    ZZ_pPush p_(p);
    ZZ pf,N,a;
    ZZ_pX g;
    Mat<ZZ> B;
    Mat<ZZ_p> C;
    Vec<ZZ_p> r;
    Vec<Pair<ZZ,long> > G,H;
    CYC s;
    
    power(pf,p,f);
    Period::init(f);
    conv(g, Period::MPoly);
    FindRoots(r,g);
    C.SetDims(1,e);
    for(i=0; i<e; i++)
        if(SolCongr(C[0], r[i]) == 0) break;

    transpose(C,C);
    kernel(C,C);
    conv(B,C);
    // optional ***************************
    // extra rows to increase probability
    // of getting shorter vector in lattice
    B.SetDims(e1<<1, e);
    for(i=0; i<e1; i++) B[i+e1] = B[i];
    for(i=0; i<e1; i++) B[i+e1][0] -= p;
    //*************************************
    j = LLL(a,B);
    a = pf;
    for(i=B.NumRows()-j; i<B.NumRows(); i++) {
        conv(q, (Period&)B[i]);
        norm(N,q);
        if(N==pf) return;
        factor(G, N/=pf);
        N = G[G.length()-1].a;
        if(N<a) { a=N; H=G; s=q; }
    }
    if(a==pf)
        throw std::runtime_error("factor not found");
    q=s;
    for(i=0; i<H.length(); i++) {
        j = CYC::ord(H[i].a);
        FactorPrime(s, H[i].a, j);
        while(H[i].b)// trial division
            if(divide(q,q,s)) H[i].b -= j;
            else conj(s,s);
    }
}

void factor(Vec<Pair<CYC, long> >& F, const CYC& a)
// F = factorization of a into cycrotomic primes
// each element of F is a pair of prime and its exponent
// such that product of prime^{exponent} is associate of a.
// Assume CYC::init(n) has been executed
// Assume n<=19 is prime.
{
    int i(0),j(0),k(0),n(CYC::n),n1(n-1),e(0),f,g;
    if(n>=23) throw std::runtime_error("n>=23 in factor");
    ZZ s,t;
    CYC b,c;
    Vec<Pair<ZZ, long> > G,H;
    F.SetLength(0);
    if(IsZero(a)) return;
    content(t,a);
    div(b,a,t);
    norm(s,b);
    while(divide(s,s,n)) e++;
    while(divide(t,t,n)) e+=n1;
    if(e) {// ramify
        F.SetLength(k+1);
        FactorPrime(F[k].a, ZZ(n));
        F[k++].b = e;
    }
    factor(G,s);
    factor(H,t);
    while(i<G.length() || j<H.length()) {
        if(j==H.length() || i<G.length() && G[i].a < H[j].a) {
            f = CYC::ord(G[i].a);
            FactorPrime(c, G[i].a, f);
            while(G[i].b) {
                if(divide(b,b,c)) {// trial division
                    F.SetLength(k+1);
                    F[k].a = c;
                    F[k].b = 0;
                    do { F[k].b++; G[i].b -= f; }
                    while(divide(b,b,c));
                    k++;
                }
                conj(c,c);
            }
            i++;
        }
        else {
            f = CYC::ord(H[j].a);
            FactorPrime(c, H[j].a, f);
            e = n1/f;
            F.SetLength(k+e);
            for(g=k; g<F.length(); g++) {
                F[g].a = c;
                F[g].b = H[j].b;
                conj(c,c);
            }
            if(i<G.length() && G[i].a == H[j].a) {
                for(g=k; G[i].b; g++) {// trial division
                    while(divide(b,b,c)) {
                        F[g].b++;
                        G[i].b -= f;
                    }
                    conj(c,c);
                }
                i++;
            }
            j++; k+=e;
        }
    }
}

void mul(CYC& a, const Vec<Pair<CYC, long> >& q)
// a = product of (cycrotomic integer)^{exponent} in q
// each element of q is a pair of integer and exponent
{
    int i;
    CYC b;
    set(a);
    for(i=0; i<q.length(); i++) {
        power(b, q[i].a, q[i].b);
        a *= b;
    }
}

void GenPrime(CYC&pi, long l, long f, long NTRY)
// generate random cyclotomic prime integer pi
// l: bit length of prime number p below pi
// f: order of p mod n
// NRY: number of trials to generate p of order f.
// Output random pi such that norm(pi)=p^f (except p=n).
// Assume CYC::init(n) has been executed.
// Assume n<=19 is prime.
// If f<=0, pi of any f is output.
{
    ZZ p;
    if(f<=0) {
        GenPrime(p,l);
        f = CYC::ord(p);
    }
    else if((CYC::n-1)%f)
        throw std::runtime_error("wrong f in GenPrime");
    else {
        long i;
        for(i=0; i<NTRY; i++) {
            GenPrime(p,l);
            if(CYC::ord(p) == f) break;
        }
        if(i==NTRY)
            throw std::runtime_error("i==NTRY in GenPrime");
    }
    FactorPrime(pi, p, f);
}

File: CYCFactoring.py
# reference:
# H.M.Edwards "Fermat's Last Theorem" sections 4.4, 4.7

from Period import *
from LLL import LLL
from linsolve import linsolve,InvMod

def SolveCongruence(u0,p):
    """ given u0 == eta_0 (mod p)
    get u_j == eta_j (mod p) (j=1,...,e-1)
    and return u_j (j=0,...,e-1)
    """
    u = -u0*Period.w[0,1:,0]
    A = Period.w[0,1:,1:].copy()
    for i in range(len(A)): A[i,i] -= u0
    return linsolve(A,u,p)

def FactorPrime(p, f=None):
    """ factoring prime number p
    p: prime number (int)
    f: order of p mod n
    return a factor pi of p such that norm(pi)==p^f.
    raise RuntimeError if factor couldn't be found.
    Assume CYC.init(n) has been executed.
    Assume n<=19 is prime.
    If f is none, f is set to ord_n(p).
    """
    if p == CYC.n: return CYC([1,-1])
    if f is None: f = CYC.ord(p)
    if f == CYC.n1: return CYC(p)
    Period.init(f)
    F = sp.factor_list(Period.MPoly, modulus=p)
    for q in F[1]:
        u0 = int(-q[0].coeff(x,0))%p
        try: u = SolveCongruence(u0,p)
        except: continue
        break

    u = (p-u) * InvMod(u0,p) % p
    a,e1 = p, Period.e-1
    for b in LLL(np.c_[u, np.eye(e1)]):
        y = Period(b)
        N = abs(y.norm())
        if N==p: return y.to_CYC()
        g = sp.factorint(N//p)
        m = max(g.keys())
        if m<a: a,h,s = m,g,y

    if a==p:
        raise RuntimeError("factor not found")
    q = s.to_CYC()
    for k in h:
        j = CYC.ord(k)
        s = FactorPrime(k,j)
        m = h[k]*f
        while m:
            t = q/s # trial division
            if t: m,q = m-j,t
            else: s = s.conj()
    return q

def factor(a):
    """ factoring cyclotomic integer a
    a: CYC object or ordinary integer
    return dictionary of {factor: exponent} pair
       such that product of factor^exponent
       is associate of a
    Assume CYC.init(n) has been executed.
    Assume n<=19 is prime.
    """
    if CYC.n>=23:
        raise RuntimeError("l>=23 in factor")
    if not isinstance(a,CYC): a = CYC(a)
    F = {}
    if a==0: return F
    d,a = a.primitive()
    G = sp.factorint(a.norm()) # primitive part
    H = sp.factorint(d) # and content
    n,n1 = CYC.n, CYC.n1

    k = G.pop(n,0) + n1*H.pop(n,0)
    if k: F[CYC([1,-1])] = k # ramify
    for k in H:
        f = CYC.ord(k)
        p = FactorPrime(k,f)
        e = n1//f
        for _ in range(e):
            F[p] = H[k]
            p = p.conj()
        if k in G:
            while G[k]:
                b = a/p # trial division
                if b:
                    F[p] += 1
                    G[k] -= f
                    a = b
                else: p = p.conj()
            G.pop(k)

    for k in G:
        f = CYC.ord(k)
        p = FactorPrime(k,f)
        while G[k]:
            b = a/p # trial division
            if b:
                if p in F: F[p] += 1
                else: F[p] = 1
                G[k] -= f
                a = b
            else: p = p.conj()

    return F

def GenPrime(l, f=None, NTRY=1000):
    """ generate random cyclotomic prime integer pi
    l: bit length of prime number p below pi
    f: order of p mod n
    NRY: number of trials to generate p of order f
    return random pi such that norm(pi)=p^f (except p=n).
    Assume CYC.init(n) has been executed.
    Assume n<=19 is prime.
    If f is None, pi of any f is returned.
    """
    if f is None:
        p = sp.randprime(1<<(l-1), 1<<l)
    elif CYC.n1%f:
        raise RuntimeError("f must divide n-1")
    else:
        for _ in range(NTRY):
            p = sp.randprime(1<<(l-1), 1<<l)
            if CYC.ord(p) == f: break
        else:
            raise RuntimeError("NTRY failed")
    return FactorPrime(p,f)
